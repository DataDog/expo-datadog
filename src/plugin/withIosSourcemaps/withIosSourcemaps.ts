/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

import type { ConfigPlugin } from "@expo/config-plugins";
import { withXcodeProject } from "@expo/config-plugins";

import { SourceMapUploadOptions } from "../getErrorTrackingPluginsFromOptions";

const SOURCEMAP_FILE_COMMAND =
  "export SOURCEMAP_FILE=$DERIVED_FILE_DIR/main.jsbundle.map";
const getDatadogXCodeCommand = (
  serviceName: SourceMapUploadOptions["serviceName"]
) =>
  `npx datadog-ci react-native xcode \`\\\"$NODE_BINARY\\\" --print \\\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\\"\`${
    serviceName ? ` --service ${serviceName}` : ""
  }`;

const withIosSourcemaps =
  (options: SourceMapUploadOptions): ConfigPlugin<void> =>
  (config) => {
    return withXcodeProject(config, async (config) => {
      const xcodeProject = config.modResults;
      const bundlePhase = xcodeProject.pbxItemByComment(
        "Bundle React Native code and images",
        "PBXShellScriptBuildPhase"
      );
      if (bundlePhase.shellScript.match("datadog-ci react-native xcode")) {
        return config;
      }
      if (bundlePhase.shellScript.match("sentry")) {
        /**
         * In the case where Sentry is already added, we add a new line to
         * bundle the javascript code twice and this time send the sourcemap
         * to Datadog.
         * That's the safest way to make sure we don't break the script.
         *
         * We also need to remove the final " and add it back for the script to
         * be well-formatted.
         */
        bundlePhase.shellScript = `${bundlePhase.shellScript.replace(
          /.$/,
          ""
        )}\\n ${getDatadogXCodeCommand(options.serviceName)}"`;

        return config;
      }

      const [beforeScript, afterScript] = bundlePhase.shellScript.split(
        "`\\\"$NODE_BINARY\\\" --print \\\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\\"`"
      );
      const datadogScript = `${SOURCEMAP_FILE_COMMAND}\\n ${getDatadogXCodeCommand(
        options.serviceName
      )}`;
      bundlePhase.shellScript = `${beforeScript}${datadogScript}${afterScript}`;

      return config;
    });
  };

export default withIosSourcemaps;
