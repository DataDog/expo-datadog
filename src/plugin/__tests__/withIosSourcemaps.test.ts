/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

import { withXcodeProject } from "@expo/config-plugins";

import withIosSourcemaps from "../withIosSourcemaps/withIosSourcemaps";
import pristineProject from "./__fixtures__/pristineProjectPbxproj.json";
import sentryProject from "./__fixtures__/sentryProjectPbxproj.json";

jest.mock("@expo/config-plugins", () => {
  return {
    ...(jest.requireActual("@expo/config-plugins") as object),
    withXcodeProject: jest.fn(),
  };
});

const mockXcodeProject = (mock: object) => {
  const xcodeProject = { ...mock };
  // @ts-ignore
  withXcodeProject.mockImplementationOnce((config, callback) => {
    // @ts-ignore
    return callback({
      ...config,
      modResults: {
        pbxItemByComment: () => xcodeProject,
      },
      xcodeProject,
    });
  });
};

/**
 * This must return a new instance each time to avoid overriding the same object
 * on different tests
 */
const createFakeConfig = () => ({
  name: "project name",
  slug: "project-name",
});

describe("withIosSourcemaps", () => {
  beforeEach(() => {
    (withXcodeProject as any).mockClear();
  });
  describe("on pristine project", () => {
    it("adds script to upload sourcemaps to Datadog", async () => {
      mockXcodeProject(pristineProject);
      const result = (await withIosSourcemaps({})(createFakeConfig())) as any;
      expect(result.xcodeProject.shellScript).toMatchInlineSnapshot(
        `"\\"if [[ -f \\\\\\"$PODS_ROOT/../.xcode.env\\\\\\" ]]; then\\\\n  source \\\\\\"$PODS_ROOT/../.xcode.env\\\\\\"\\\\nfi\\\\nif [[ -f \\\\\\"$PODS_ROOT/../.xcode.env.local\\\\\\" ]]; then\\\\n  source \\\\\\"$PODS_ROOT/../.xcode.env.local\\\\\\"\\\\nfi\\\\n\\\\n# The project root by default is one level up from the ios directory\\\\nexport PROJECT_ROOT=\\\\\\"$PROJECT_DIR\\\\\\"/..\\\\n\\\\nif [[ \\\\\\"$CONFIGURATION\\\\\\" = *Debug* ]]; then\\\\n  export SKIP_BUNDLING=1\\\\nfi\\\\nexport SOURCEMAP_FILE=$DERIVED_FILE_DIR/main.jsbundle.map\\\\n npx datadog-ci react-native xcode \`\\\\\\"$NODE_BINARY\\\\\\" --print \\\\\\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\\\\\"\`\\\\n\\\\n\\""`
      );
    });
    it("adds script to upload sourcemaps to Datadog with custom service name", async () => {
      mockXcodeProject(pristineProject);
      const result = (await withIosSourcemaps({
        serviceName: "com.company.app",
      })(createFakeConfig())) as any;
      expect(result.xcodeProject.shellScript).toMatchInlineSnapshot(
        `"\\"if [[ -f \\\\\\"$PODS_ROOT/../.xcode.env\\\\\\" ]]; then\\\\n  source \\\\\\"$PODS_ROOT/../.xcode.env\\\\\\"\\\\nfi\\\\nif [[ -f \\\\\\"$PODS_ROOT/../.xcode.env.local\\\\\\" ]]; then\\\\n  source \\\\\\"$PODS_ROOT/../.xcode.env.local\\\\\\"\\\\nfi\\\\n\\\\n# The project root by default is one level up from the ios directory\\\\nexport PROJECT_ROOT=\\\\\\"$PROJECT_DIR\\\\\\"/..\\\\n\\\\nif [[ \\\\\\"$CONFIGURATION\\\\\\" = *Debug* ]]; then\\\\n  export SKIP_BUNDLING=1\\\\nfi\\\\nexport SOURCEMAP_FILE=$DERIVED_FILE_DIR/main.jsbundle.map\\\\n npx datadog-ci react-native xcode \`\\\\\\"$NODE_BINARY\\\\\\" --print \\\\\\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\\\\\"\` --service com.company.app\\\\n\\\\n\\""`
      );
    });
  });
  describe("on projects implementing Sentry", () => {
    it("adds script to upload sourcemaps to Datadog", async () => {
      mockXcodeProject(sentryProject);
      const result = (await withIosSourcemaps({})(createFakeConfig())) as any;
      expect(result.xcodeProject.shellScript).toMatchInlineSnapshot(
        `"\\"export SENTRY_PROPERTIES=sentry.properties\\\\nexport EXTRA_PACKAGER_ARGS=\\\\\\"--sourcemap-output $DERIVED_FILE_DIR/main.jsbundle.map\\\\\\"\\\\nif [[ -f \\\\\\"$PODS_ROOT/../.xcode.env\\\\\\" ]]; then\\\\n  source \\\\\\"$PODS_ROOT/../.xcode.env\\\\\\"\\\\nfi\\\\nif [[ -f \\\\\\"$PODS_ROOT/../.xcode.env.local\\\\\\" ]]; then\\\\n  source \\\\\\"$PODS_ROOT/../.xcode.env.local\\\\\\"\\\\nfi\\\\n\\\\n# The project root by default is one level up from the ios directory\\\\nexport PROJECT_ROOT=\\\\\\"$PROJECT_DIR\\\\\\"/..\\\\n\\\\nif [[ \\\\\\"$CONFIGURATION\\\\\\" = *Debug* ]]; then\\\\n  export SKIP_BUNDLING=1\\\\nfi\\\\n\`node --print \\\\\\"require.resolve('@sentry/cli/package.json').slice(0, -13) + '/bin/sentry-cli'\\\\\\"\` react-native xcode --force-foreground \`\\\\\\"$NODE_BINARY\\\\\\" --print \\\\\\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\\\\\"\`\\\\n\\\\n\\\\n npx datadog-ci react-native xcode \`\\\\\\"$NODE_BINARY\\\\\\" --print \\\\\\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\\\\\"\`\\""`
      );
    });
    it("adds script to upload sourcemaps to Datadog with custom service name", async () => {
      mockXcodeProject(sentryProject);
      const result = (await withIosSourcemaps({
        serviceName: "com.company.app",
      })(createFakeConfig())) as any;
      expect(result.xcodeProject.shellScript).toMatchInlineSnapshot(
        `"\\"export SENTRY_PROPERTIES=sentry.properties\\\\nexport EXTRA_PACKAGER_ARGS=\\\\\\"--sourcemap-output $DERIVED_FILE_DIR/main.jsbundle.map\\\\\\"\\\\nif [[ -f \\\\\\"$PODS_ROOT/../.xcode.env\\\\\\" ]]; then\\\\n  source \\\\\\"$PODS_ROOT/../.xcode.env\\\\\\"\\\\nfi\\\\nif [[ -f \\\\\\"$PODS_ROOT/../.xcode.env.local\\\\\\" ]]; then\\\\n  source \\\\\\"$PODS_ROOT/../.xcode.env.local\\\\\\"\\\\nfi\\\\n\\\\n# The project root by default is one level up from the ios directory\\\\nexport PROJECT_ROOT=\\\\\\"$PROJECT_DIR\\\\\\"/..\\\\n\\\\nif [[ \\\\\\"$CONFIGURATION\\\\\\" = *Debug* ]]; then\\\\n  export SKIP_BUNDLING=1\\\\nfi\\\\n\`node --print \\\\\\"require.resolve('@sentry/cli/package.json').slice(0, -13) + '/bin/sentry-cli'\\\\\\"\` react-native xcode --force-foreground \`\\\\\\"$NODE_BINARY\\\\\\" --print \\\\\\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\\\\\"\`\\\\n\\\\n\\\\n npx datadog-ci react-native xcode \`\\\\\\"$NODE_BINARY\\\\\\" --print \\\\\\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\\\\\"\` --service com.company.app\\""`
      );
    });
  });
});
