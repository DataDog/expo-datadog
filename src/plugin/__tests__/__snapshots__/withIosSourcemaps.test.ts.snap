// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withIosSourcemaps on pristine project adds script to upload sourcemaps to Datadog 1`] = `""if [[ -f \\"$PODS_ROOT/../.xcode.env\\" ]]; then\\n  source \\"$PODS_ROOT/../.xcode.env\\"\\nfi\\nif [[ -f \\"$PODS_ROOT/../.xcode.env.local\\" ]]; then\\n  source \\"$PODS_ROOT/../.xcode.env.local\\"\\nfi\\n\\n# The project root by default is one level up from the ios directory\\nexport PROJECT_ROOT=\\"$PROJECT_DIR\\"/..\\n\\nif [[ \\"$CONFIGURATION\\" = *Debug* ]]; then\\n  export SKIP_BUNDLING=1\\nfi\\nexport SOURCEMAP_FILE=$DERIVED_FILE_DIR/main.jsbundle.map\\n \\n\\nif [[ -z \\"$DATADOG_CI_EXEC\\" ]]; then\\n    DATADOG_CI_EXEC=\\"$(\\"$NODE_BINARY\\" --print \\"require('path').resolve(require('path').dirname(require.resolve('@datadog/datadog-ci/package.json')), '../../.bin/datadog-ci')\\")\\";\\n    \\n    # Check if the file exists and is executable\\n    if [[ -x \\"$DATADOG_CI_EXEC\\" ]]; then\\n        export DATADOG_CI_EXEC;\\n    else\\n        echo \\"Error: DATADOG_CI_EXEC does not exist or is not executable\\";\\n        exit 1;\\n    fi\\nfi\\n\\n\\n$DATADOG_CI_EXEC react-native xcode \`\\"$NODE_BINARY\\" --print \\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\"\`\\n\\n""`;

exports[`withIosSourcemaps on pristine project adds script to upload sourcemaps to Datadog with custom service name 1`] = `""if [[ -f \\"$PODS_ROOT/../.xcode.env\\" ]]; then\\n  source \\"$PODS_ROOT/../.xcode.env\\"\\nfi\\nif [[ -f \\"$PODS_ROOT/../.xcode.env.local\\" ]]; then\\n  source \\"$PODS_ROOT/../.xcode.env.local\\"\\nfi\\n\\n# The project root by default is one level up from the ios directory\\nexport PROJECT_ROOT=\\"$PROJECT_DIR\\"/..\\n\\nif [[ \\"$CONFIGURATION\\" = *Debug* ]]; then\\n  export SKIP_BUNDLING=1\\nfi\\nexport SOURCEMAP_FILE=$DERIVED_FILE_DIR/main.jsbundle.map\\n \\n\\nif [[ -z \\"$DATADOG_CI_EXEC\\" ]]; then\\n    DATADOG_CI_EXEC=\\"$(\\"$NODE_BINARY\\" --print \\"require('path').resolve(require('path').dirname(require.resolve('@datadog/datadog-ci/package.json')), '../../.bin/datadog-ci')\\")\\";\\n    \\n    # Check if the file exists and is executable\\n    if [[ -x \\"$DATADOG_CI_EXEC\\" ]]; then\\n        export DATADOG_CI_EXEC;\\n    else\\n        echo \\"Error: DATADOG_CI_EXEC does not exist or is not executable\\";\\n        exit 1;\\n    fi\\nfi\\n\\n\\n$DATADOG_CI_EXEC react-native xcode \`\\"$NODE_BINARY\\" --print \\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\"\` --service com.company.app\\n\\n""`;

exports[`withIosSourcemaps on projects implementing Sentry adds script to upload sourcemaps to Datadog 1`] = `""export SENTRY_PROPERTIES=sentry.properties\\nexport EXTRA_PACKAGER_ARGS=\\"--sourcemap-output $DERIVED_FILE_DIR/main.jsbundle.map\\"\\nif [[ -f \\"$PODS_ROOT/../.xcode.env\\" ]]; then\\n  source \\"$PODS_ROOT/../.xcode.env\\"\\nfi\\nif [[ -f \\"$PODS_ROOT/../.xcode.env.local\\" ]]; then\\n  source \\"$PODS_ROOT/../.xcode.env.local\\"\\nfi\\n\\n# The project root by default is one level up from the ios directory\\nexport PROJECT_ROOT=\\"$PROJECT_DIR\\"/..\\n\\nif [[ \\"$CONFIGURATION\\" = *Debug* ]]; then\\n  export SKIP_BUNDLING=1\\nfi\\n\`node --print \\"require.resolve('@sentry/cli/package.json').slice(0, -13) + '/bin/sentry-cli'\\"\` react-native xcode --force-foreground \`\\"$NODE_BINARY\\" --print \\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\"\`\\n\\n\\n \\n\\nif [[ -z \\"$DATADOG_CI_EXEC\\" ]]; then\\n    DATADOG_CI_EXEC=\\"$(\\"$NODE_BINARY\\" --print \\"require('path').resolve(require('path').dirname(require.resolve('@datadog/datadog-ci/package.json')), '../../.bin/datadog-ci')\\")\\";\\n    \\n    # Check if the file exists and is executable\\n    if [[ -x \\"$DATADOG_CI_EXEC\\" ]]; then\\n        export DATADOG_CI_EXEC;\\n    else\\n        echo \\"Error: DATADOG_CI_EXEC does not exist or is not executable\\";\\n        exit 1;\\n    fi\\nfi\\n\\n\\n$DATADOG_CI_EXEC react-native xcode \`\\"$NODE_BINARY\\" --print \\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\"\`""`;

exports[`withIosSourcemaps on projects implementing Sentry adds script to upload sourcemaps to Datadog with custom service name 1`] = `""export SENTRY_PROPERTIES=sentry.properties\\nexport EXTRA_PACKAGER_ARGS=\\"--sourcemap-output $DERIVED_FILE_DIR/main.jsbundle.map\\"\\nif [[ -f \\"$PODS_ROOT/../.xcode.env\\" ]]; then\\n  source \\"$PODS_ROOT/../.xcode.env\\"\\nfi\\nif [[ -f \\"$PODS_ROOT/../.xcode.env.local\\" ]]; then\\n  source \\"$PODS_ROOT/../.xcode.env.local\\"\\nfi\\n\\n# The project root by default is one level up from the ios directory\\nexport PROJECT_ROOT=\\"$PROJECT_DIR\\"/..\\n\\nif [[ \\"$CONFIGURATION\\" = *Debug* ]]; then\\n  export SKIP_BUNDLING=1\\nfi\\n\`node --print \\"require.resolve('@sentry/cli/package.json').slice(0, -13) + '/bin/sentry-cli'\\"\` react-native xcode --force-foreground \`\\"$NODE_BINARY\\" --print \\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\"\`\\n\\n\\n \\n\\nif [[ -z \\"$DATADOG_CI_EXEC\\" ]]; then\\n    DATADOG_CI_EXEC=\\"$(\\"$NODE_BINARY\\" --print \\"require('path').resolve(require('path').dirname(require.resolve('@datadog/datadog-ci/package.json')), '../../.bin/datadog-ci')\\")\\";\\n    \\n    # Check if the file exists and is executable\\n    if [[ -x \\"$DATADOG_CI_EXEC\\" ]]; then\\n        export DATADOG_CI_EXEC;\\n    else\\n        echo \\"Error: DATADOG_CI_EXEC does not exist or is not executable\\";\\n        exit 1;\\n    fi\\nfi\\n\\n\\n$DATADOG_CI_EXEC react-native xcode \`\\"$NODE_BINARY\\" --print \\"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\\"\` --service com.company.app""`;
